Hw12 web server

daria-grebneva opened the pull request.

daria-grebneva committed:
Начала работу над веб сервером

daria-grebneva committed:
Добавила получение списка клинтов и форму для добавления нового клиента

stvort commented on hw12-web-server/src/main/java/ru/otus/WebServerWithFilterBasedSecurityDemo.java:
```diff
@@ -0,0 +1,77 @@
+    // REST сервис
+    http://localhost:8080/api/clients
+*/
+public class WebServerWithFilterBasedSecurityDemo {
```
Это не демо веб-сервера, это приложение по работе с клиентами

daria-grebneva replied:
Переименовала

stvort commented on hw12-web-server/src/main/java/ru/otus/WebServerWithFilterBasedSecurityDemo.java:
```diff
@@ -0,0 +1,77 @@
+        var clientDao = new ClientDaoImpl(transactionManager, clientTemplate);
+
+
+        Client client1 = new Client(1L, "Daria", new Address(1L, "Sovetskaya 116"), Collections.singletonList(new Phone(1L, "+79111234565")));
+        Client client2 = new Client(2L, "Maria", new Address(2L, "Sovetskaya 16"), Collections.singletonList(new Phone(2L, "+79231234515")));
+        clientDao.saveClient(client1);
+        clientDao.saveClient(client2);
```
Это лучше сделать через миграции

daria-grebneva replied:
Поправлено

stvort commented on hw12-web-server/src/main/java/ru/otus/WebServerWithFilterBasedSecurityDemo.java:
```diff
@@ -0,0 +1,77 @@
+
+        var clientTemplate = new DataTemplateHibernate<>(Client.class);
+
+        var clientDao = new ClientDaoImpl(transactionManager, clientTemplate);
```
Как я говорил на занятии, **не должно быть никаких дао клиентов**, все абстракции по работе с ними должны быть от ДЗ по гиберейту, а там у нас сервис

daria-grebneva replied:
Переименовала обратно в сервис

stvort commented on hw12-web-server/src/main/java/ru/otus/server/UsersWebServerSimple.java:
```diff
@@ -0,0 +1,84 @@
+    private static final String COMMON_RESOURCES_DIR = "static";
+
+    private final ClientDao clientDao;
+    private final UserDao userDao;
```
Это не используется

daria-grebneva replied:
Убрала неиспользуемый userDao

stvort commented on hw12-web-server/src/main/java/ru/otus/server/UsersWebServerSimple.java:
```diff
@@ -0,0 +1,84 @@
+import ru.otus.servlet.*;
+
+
+public class UsersWebServerSimple implements UsersWebServer {
```
Это не сервер юзеров, это сервер приложения по работе с клиентами. Мы говорили на занятии, что в ДЗ можно вообще обойтись без юзеров или сделать одного-двух предустановленных. Так или иначе называть сервер в их честь, избыточно

daria-grebneva replied:
Поправила

stvort commented on hw12-web-server/src/main/java/ru/otus/server/UsersWebServerSimple.java:
```diff
@@ -0,0 +1,84 @@
+
+        HandlerList handlers = new HandlerList();
+        handlers.addHandler(resourceHandler);
+        handlers.addHandler(applySecurity(servletContextHandler, "/users", "/api/user/*", "/api/clients", "/client"));
```
Как и держать для них url в коде ("/users") для которого все равно нет сервлета

daria-grebneva replied:
Поправлено

stvort commented on hw12-web-server/src/main/resources/templates/users.html:
```diff
@@ -0,0 +1,43 @@
+<html xmlns="http://www.w3.org/1999/xhtml">
```
Да и страница юзерам не нужна

daria-grebneva replied:
Удалила ненужную страницу

stvort commented on hw12-web-server/src/main/java/ru/otus/servlet/ClientsApiServlet.java:
```diff
@@ -0,0 +1,62 @@
+    protected void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+        String action = request.getParameter("action");
+
+        if ("submit".equals(action)) {
+            Long id = getIdOfNewFields();
+            Address address = new Address(id, request.getParameter("clientAddress"));
+            Phone phone = new Phone(id, request.getParameter("clientPhone"));
+            Client client = new Client(id, request.getParameter("clientName"), address, Collections.singletonList(phone));
+            try {
+                clientDao.saveClient(client);
+            } catch (Exception e) {
+                response.setStatus(SC_INTERNAL_SERVER_ERROR);
+            }
+            response.setStatus(SC_OK);
+            response.sendRedirect("/client");
```
Не совсем. Как обсуждалось на занятии, API это про данные, клиентом API может быть кто угодно, от веб-приложений до мобильных и десктоп приложений и консольных утилит. Многим из них вообще не интересны какие-то веб страницы. Поэтому, в API мы принимаем и отдаем данные (например, в JSON). Причем данные классических форм, API не обрабатывает, этим пусть занимается обычный контроллер (в нашем случае сервлет), результатом обработки которого должна стать веб страница (или редирект). Данные же для API обычно отправляются через какой-либо ЯП, в нашем случае это JS

daria-grebneva replied:
Спасибо большое, я поправила) Когда дело доходит до практики, воспринимается совсем иначе))

daria-grebneva committed:
Правки комментариев

stvort commented on hw12-web-server/src/main/java/ru/otus/WebServerWithFilterBasedSecurityRunner.java:
```diff
@@ -0,0 +1,68 @@
+
+        var clientTemplate = new DataTemplateHibernate<>(Client.class);
+
+        var clientDao = new DBServiceClientImpl(transactionManager, clientTemplate);
```
Дао это дао, сервис это сервис. Это разные вещи, которые находятся на разных слоях многослойной архитектуры. У нас тут сервис

daria-grebneva replied:
Поправила именование

stvort commented on hw12-web-server/src/main/resources/static/index.html:
```diff
@@ -0,0 +1,13 @@
+    <img src="img/logo.svg" style="width: 300px; height: 300px">
+    <br/>
+    <h3>Добро пожаловать на стартовую страницу примера</h3>
+    <a href="clients">Перейти на странцу информации о пользователях (возможно защищено паролем)</a>
```
Каких пользователях?

daria-grebneva replied:
Поправила

stvort commented on hw12-web-server/src/main/resources/static/index.html:
```diff
@@ -0,0 +1,13 @@
+    <img src="img/logo.svg" style="width: 300px; height: 300px">
+    <br/>
+    <h3>Добро пожаловать на стартовую страницу примера</h3>
+    <a href="clients">Перейти на странцу информации о пользователях (возможно защищено паролем)</a>
```
А у нас нет такого урла "clients", "client" есть, "/api/clients" есть, а вот "clients" нет. Приложение не работает

daria-grebneva replied:
Поправлено

stvort commented on hw12-web-server/src/main/resources/db/migration/V2__clients.sql:
```diff
@@ -0,0 +1,7 @@
+insert into address (street, id) values ('Sovetskaya 116', 1);
+insert into client (name, id) values ('Daria', 1);
```
Так не идет последовательность. И когда мы сохраняем клиента через приложение, то секвенс выдает при вставке id = 1, а такой уже есть

daria-grebneva replied:
Поменяла на корректную работу с последовательностью

daria-grebneva committed:
Внесла правки по комментариям:
   - Корректная работа в миграциях с последовательностями
   - Переименования
   - Поправила url "clients" -> "client"

stvort commented on hw12-web-server/src/main/java/ru/otus/servlet/ClientsServlet.java:
```diff
@@ -0,0 +1,54 @@
+
+    @Override
+    protected void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+        String action = request.getParameter("action");
+
+        if ("submit".equals(action)) {
```
Зачем это?

daria-grebneva replied:
При предыдущих попытках реализовать пост запрос не работало без флага, еще при первых версиях дз. Сейчас это излишне, убрала, спасибо!

stvort commented on hw12-web-server/src/main/java/ru/otus/servlet/ClientsServlet.java:
```diff
@@ -0,0 +1,54 @@
+        }
+    }
+
+    private Long getIdOfNewFields() {
+        List<Client> clients = dbServiceClient.findAll();
+        return (long) clients.size() + 1L;
+    }
```
Это просто не будет работать в реальности. Я создал трех клиентов в БД. Потом первого удалил. Остались id 2 и 3. Значит при текущем подходе id будет равен 3, а такой уже есть. Все же пусть лучше Id берется при вставке согласно стратегии

daria-grebneva replied:
Добавила новые конструкторы в модели, чтобы id генерился сам и выпилила эту странную логику

daria-grebneva committed:
Правки по комментариям:
  - убрала странную логику для получения id
  - убрала ненужный параметр в post запросе

daria-grebneva committed:
Merge branch 'main' into hw12-web-server
